File: dgv undo~redo/[C#]-dgv undo~redo/C#/dgv undo~redo cs version/extensions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace WindowsFormsApplication1
{
    static class extensions
    {
        public static bool NeedsItem(this Stack<object[][]> instance, DataGridView dgv)
        {
            if (instance.Count == 0)
            {
                return true;
            }
            object[][] rows = instance.Peek();
            return !ItemEquals(rows, dgv.Rows.Cast<DataGridViewRow>().Where(r => !r.IsNewRow).ToArray());
        }

        public static bool ItemEquals(this object[][] instance, DataGridViewRow[] dgvRows)
        {
            if (instance.Count() != dgvRows.Count())
            {
                return false;
            }
            return !Enumerable.Range(0, instance.GetLength(0)).Any(x => !instance[x].SequenceEqual(dgvRows[x].Cells.Cast<DataGridViewCell>().Select(c => c.Value).ToArray()));
        }

    }

}

File: dgv undo~redo/[VB]-dgv undo~redo/VB/dgv undo~redo/extensions.vb
Module extensions
    <System.Runtime.CompilerServices.Extension()> _
    Public Function NeedsItem(ByVal instance As Stack(Of Object()()), ByVal dgv As DataGridView) As Boolean
        If instance.Count = 0 Then Return True
        Dim rows()() As Object = instance.Peek
        Return Not ItemEquals(rows, dgv.Rows.Cast(Of DataGridViewRow).Where(Function(r) Not r.IsNewRow).ToArray)
    End Function

    <System.Runtime.CompilerServices.Extension()> _
    Public Function ItemEquals(ByVal instance As Object()(), ByVal dgvRows() As DataGridViewRow) As Boolean
        If instance.Count <> dgvRows.Count Then Return False
        Return Not Enumerable.Range(0, instance.GetLength(0)).Any(Function(x) Not instance(x).SequenceEqual(dgvRows(x).Cells.Cast(Of DataGridViewCell).Select(Function(c) c.Value).ToArray))
    End Function

End Module
